disttask/framework/scheduler/scheduler.go

```diff
@@ -0,0 +1,284 @@
+func NewInternalScheduler(ctx context.Context, id string, taskID int64, subtaskTable SubtaskTable, pool Pool) InternalScheduler {
+	logPrefix := fmt.Sprintf("id: %s, task_id: %d", id, taskID)
+	schedulerImpl := &InternalSchedulerImpl{
+		id:           id,
+		taskID:       taskID,
+		subtaskTable: subtaskTable,
+		pool:         pool,
+		logCtx:       logutil.WithKeyValue(context.Background(), "scheduler", logPrefix),
+	}
+	schedulerImpl.ctx, schedulerImpl.cancel = context.WithCancel(ctx)
+
+	return schedulerImpl
+}
+
+// Start starts the scheduler.
+func (*InternalSchedulerImpl) Start() {
+	//	s.wg.Add(1)
+	//	go func() {
+	//		defer s.wg.Done()
+	//		s.heartbeat()
+	//	}()
+}
+
+// Stop stops the scheduler.
+func (s *InternalSchedulerImpl) Stop() {
+	s.cancel()
+	s.wg.Wait()
+}
+
+//	func (s *InternalSchedulerImpl) heartbeat() {
+//		ticker := time.NewTicker(proto.HeartbeatInterval)
+//		for {
+//			select {
+//			case <-s.ctx.Done():
+//				return
+//			case <-ticker.C:
+//				if err := s.subtaskTable.UpdateHeartbeat(s.id, s.taskID, time.Now()); err != nil {
+//					s.onError(err)
+//					return
+//				}
+//			}
+//		}
+//	}
+
+// Run runs the scheduler task.
+func (s *InternalSchedulerImpl) Run(ctx context.Context, task *proto.Task) error {
+	runCtx, runCancel := context.WithCancel(ctx)
+	defer runCancel()
+	s.registerCancelFunc(runCancel)
+
+	s.resetError()
+	logutil.Logger(s.logCtx).Info("scheduler run a step", zap.Any("step", task.Step), zap.Any("concurrency", task.Concurrency))
+	scheduler, err := createScheduler(task)
+	if err != nil {
+		s.onError(err)
+		return s.getError()
+	}
+
+	if err := scheduler.InitSubtaskExecEnv(runCtx); err != nil {
+		s.onError(err)
+		return s.getError()
+	}
+	defer func() {
+		err := scheduler.CleanupSubtaskExecEnv(runCtx)
+		if err != nil {
+			logutil.Logger(s.logCtx).Error("cleanup subtask exec env failed", zap.Error(err))
+		}
+	}()
+
+	minimalTaskCh := make(chan func(), task.Concurrency)
+	defer close(minimalTaskCh)
+	var minimalTaskWg sync.WaitGroup
+
+	err = s.pool.RunWithConcurrency(minimalTaskCh, uint32(task.Concurrency))
+	if err != nil {
+		s.onError(err)
+		return s.getError()
+	}
+
+	for {
+		subtask, err := s.subtaskTable.GetSubtaskInStates(s.id, task.ID, proto.TaskStatePending)
+		if err != nil {
+			s.onError(err)
+			break
+		}
+		if subtask == nil {
+			logutil.Logger(s.logCtx).Info("scheduler finished subtasks", zap.Any("step", task.Step))
+			break
+		}
+		s.updateSubtaskState(subtask.ID, proto.TaskStateRunning)
+		if err := s.getError(); err != nil {
+			break
+		}
+
+		minimalTasks := scheduler.SplitSubtask(subtask.Meta)
+		logutil.Logger(s.logCtx).Info("split subTask", zap.Any("cnt", len(minimalTasks)), zap.Any("subtask_id", subtask.ID))
+		for _, minimalTask := range minimalTasks {
+			minimalTaskWg.Add(1)
+			j := minimalTask
+			minimalTaskCh <- func() {
+				s.runMinimalTask(runCtx, j, task.Type, task.Step)
+				minimalTaskWg.Done()
+			}
+		}
+		minimalTaskWg.Wait()
```

Comment:
It seems that two subtasks from one global task runs in serial. Is there a TODO to run them in concurrency in the future?